### DISCUSSION TITLE
`assert` statement

### DISCUSSION THREAD

**Discussion by vadimkantorov** [dotnet/csharplang#7502] [Language Ideas]
Must have been discussed before, but I could not find it :( Maybe because my search queries were not specific enough. Opening it in Discussions, hoping this does not clutter things too much :(

It's nice to be able to write assertions more concise than `System.Diagnostics.Debug.Assert(condition)` and to be able to not execute the code under `condition` given some release (or release-like) compilation flag.

This kind of statement exists in many other languages as native feature.

**Comment by HaloFour** [comment]
You can accomplish this today with global usings, which can be applied project-wide:

```cs
global using static System.Diagnostics.Debug;
```

Then you can just use `Assert` without the `Debug` class name:

```cs
void M(string s) {
    Assert(s != null, "s shouldn't be null!");
}
```

The method is already marked with `ConditionalAttribute` which causes the method call to be elided if you do a Release build.

**Reply by vadimkantorov** [reply]
> You can accomplish this today with global usings, which can be applied project-wide:

I agree, this helps! But might worth somehow enabling it by default or creating an alternative language-supported feature to make it even more wide-spread / recommended / concise. To promote use of these keywords. Maybe worth making it generic to directly throw some custom exception? E.g. there's plenty of if(cond) throw new SomeException(some_message), many times these would be much simpler to read if written as assert<SomeException> cond, some_message (if somehow python style) or assert cond, throw new SomeException(some_message)

Such asserts are also useful for pedagogical/teaching code (so not requiring a special file for global usings is nice)

**Reply by huoyaoyuan** [reply]
> But might worth somehow enabling it by default or creating an alternative language-supported feature to make it even more wide-spread / recommended / concise.

Library features are preferred over language built-ins.

> Maybe worth making it generic to directly throw some custom exception?

You can already define your own one, and that's how `Assert` in unit test frameworks work.

**Reply by huoyaoyuan** [reply]
Also note that `ConditionalAttribute` elides the parameter evaluation too. It just works like `#ifdef`.
